sharktank/sharktank/models/clip/clip.py

```diff
@@ -0,0 +1,502 @@
+# Copyright 2021 The OpenAI Team Authors and The HuggingFace Team.
+# Copyright 2024 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Any, Optional, Tuple, Union
+import torch
+from torch import nn
+import transformers
+
+# TODO: port _prepare_4d_attention_mask and _create_4d_causal_attention_mask to sharktank
+from transformers.modeling_attn_mask_utils import (
+    _prepare_4d_attention_mask,
+    _create_4d_causal_attention_mask,
+)
+from collections import OrderedDict
+
+from ...layers import BaseLayer, LinearLayer, LayerNorm, TokenEmbeddingLayer
+from ... import ops
+from ...types.theta import Theta, Dataset
+from ...types.tensors import DefaultPrimitiveTensor
+from ...layers.configs import ClipTextConfig
+from ...layers.activations import ACT2FN
+
+
+class ClipTextEmbeddings(nn.Module):
+    def __init__(self, theta: Theta, config: ClipTextConfig):
+        super().__init__()
+        self.token_embedding = TokenEmbeddingLayer(
+            theta=theta("token_embedding"), dtype=None
+        )
+        self.position_embedding = TokenEmbeddingLayer(
+            theta=theta("position_embedding"), dtype=None
+        )
+
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer(
+            "position_ids",
+            torch.arange(config.max_position_embeddings).expand((1, -1)),
+            persistent=False,
+        )
+
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+    ) -> torch.Tensor:
+        seq_length = (
+            input_ids.shape[-1] if input_ids is not None else inputs_embeds.shape[-2]
+        )
+
+        if position_ids is None:
+            position_ids = self.position_ids[:, :seq_length]
+
+        if inputs_embeds is None:
+            inputs_embeds = self.token_embedding(input_ids)
+
+        position_embeddings = self.position_embedding(position_ids)
+        embeddings = inputs_embeds + position_embeddings
+
+        return embeddings
+
+
+class ClipAttention(BaseLayer):
```

Comment:
At a quick glance is seems much of this is reused from the punet Attention (outside the decomposed SDPA which we should try avoiding) from here https://github.com/nod-ai/shark-ai/blob/5f98e7321564eeec6d0901250699938c7307ef65/sharktank/sharktank/models/punet/layers.py#L390. Can we just extend and reuse that implementation?

Reply:
I would like to iterate on improving code reuse and performance in subsequent PRs.